### ISSUE TITLE
Do not expose students to variants with mustache errors

### ISSUE THREAD

**Issue by ppannuto** [PrairieLearn/PrairieLearn#8859]
Following up from this [brief slack discussion](https://prairielearn.slack.com/archives/C266KEH9A/p1700155897527049). It would be really valuable to have an option that hard-failed variants with `mustache` errors.

Most specifically to the immediate need, if `mustache` tries to reference a variable that doesn't exist, we would like PL to follow the 'broken variant' flow or similar, i.e. logging an error and not displaying that variant to the student at all, simply generate another one.

In the immediate term, this is a workaround to the 'stale variant' issue. More generally, however, I can't really think of a situation where a template with a bug/error/warning of any kind _should_ be displayed to students. If there's a template error, that means there's an error with the question we need to fix! I would expect the behavior to be the same as in `server.py`, where variants which generate errors are not shown to students.

**Comment by nwalters512** [comment]
This will require https://github.com/janl/mustache.js/issues/599, which unfortunately seems unlikely to resolve. There are some other potential solutions mentioned on that issue, including some that recursively wrap data in proxy objects or switching to something like handlebars.

Regardless of the approach to be taken here, this will require a careful rollout in order to avoid impacting existing questions. I imagine this can be opt-in on a per-course basis to start with.

**Comment by eliotwrobson** [comment]
Looking at this again, it might be interesting to have this on a _per-question_ basis. I could see myself enabling this just for debugging a question while I'm writing it.

**Comment by eliotwrobson** [comment]
> I see that someone made https://www.npmjs.com/package/mustache-validator which might do exactly what we want?

This can be done with Chevron as well: https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/elements/pl-template/pl-template.py#L141

We don't even have to switch to Chevron rendering to use it, just discard the rendering output and emit the error messages.

> I agree with @/ppannuto that we really should be hard-erroring on template errors.

@mwest1066 I think these errors would be very well suited for https://github.com/PrairieLearn/PrairieLearn/pull/7983

**Comment by mwest1066** [comment]
That's a great observation that `chevron` supports erroring if data is missing. If we implemented this then I could just about be talked into enabling it unconditionally.

It's also an interesting idea about using instructor-only errors for this. I'm not sure whether I think it's better or worse to show these errors to students as well, just like we do with other errors?

**Comment by eliotwrobson** [comment]
> That's a great observation that `chevron` supports erroring if data is missing. If we implemented this then I could just about be talked into enabling it unconditionally.
> 
> It's also an interesting idea about using instructor-only errors for this. I'm not sure whether I think it's better or worse to show these errors to students as well, just like we do with other errors?

I could be wrong about this, but isn't the style of writing mustache templates to have certain sections ignored by leaving out variables? A lot of the element code rendering does this (with Chevron), and I imagine a lot of question code does as well. Enabling errors on missing keys is a significant departure from this style, and means that all potential keys have to have blank values added. There was discussion about this convention when we were working on the `pl-template` element: https://github.com/PrairieLearn/PrairieLearn/pull/6747#issuecomment-1368272225

**Comment by mwest1066** [comment]
> I could be wrong about this, but isn't the style of writing mustache templates to have certain sections ignored by leaving out variables? A lot of the element code rendering does this (with Chevron), and I imagine a lot of question code does as well. Enabling errors on missing keys is a significant departure from this style, and means that all potential keys have to have blank values added. There was discussion about this convention when we were working on the `pl-template` element: [#6747 (comment)](https://github.com/PrairieLearn/PrairieLearn/pull/6747#issuecomment-1368272225)

Yes, you are correct. This is definitely a problem for the approach articulated here. Is this why you were thinking that instructor-only warnings would be the right way to handle this?

**Comment by reteps** [comment]
That thread provided an idea of how to catch those bugs: https://github.com/janl/mustache.js/issues/599#issuecomment-2691782796
